python/src/trezorlib/cardano.py

```diff
@@ -315,6 +321,69 @@ def create_withdrawal(withdrawal) -> messages.CardanoTxWithdrawalType:
     )
 
 
+def create_auxiliary_data(auxiliary_data) -> messages.CardanoTxAuxiliaryDataType:
+    if auxiliary_data is None:
+        return None
+
+    AUXILIARY_DATA_MISSING_FIELDS_ERROR = ValueError(
+        "Auxiliary data is missing some fields"
+    )
+
+    if "type" not in auxiliary_data:
+        raise AUXILIARY_DATA_MISSING_FIELDS_ERROR
+
+    auxiliary_data_type = auxiliary_data["type"]
+    if auxiliary_data_type == messages.CardanoAuxiliaryDataType.BLOB:
+        if "blob" not in auxiliary_data:
+            raise AUXILIARY_DATA_MISSING_FIELDS_ERROR
+
+        return messages.CardanoTxAuxiliaryDataType(
+            type=auxiliary_data_type,
+            blob=bytes.fromhex(auxiliary_data["blob"]),
+        )
+    elif auxiliary_data_type == messages.CardanoAuxiliaryDataType.TUPLE:
+        if "metadata" not in auxiliary_data:
+            raise AUXILIARY_DATA_MISSING_FIELDS_ERROR
+
+        metadata = auxiliary_data["metadata"]
+        if "type" not in metadata:
+            raise AUXILIARY_DATA_MISSING_FIELDS_ERROR
+
+        metadata_type = metadata["type"]
+        if metadata_type == messages.CardanoMetadataType.CATALYST_REGISTRATION:
+            catalyst_registration = metadata["catalyst_registration_parameters"]
+            if not all(
+                k in catalyst_registration
+                for k in REQUIRED_FIELDS_CATALYST_REGISTRATION
+            ):
+                raise AUXILIARY_DATA_MISSING_FIELDS_ERROR
+
+            catalyst_registration_parameters = (
+                messages.CardanoCatalystRegistrationParametersType(
+                    voting_public_key=bytes.fromhex(
+                        catalyst_registration["voting_public_key"]
+                    ),
+                    staking_path=tools.parse_path(
+                        catalyst_registration["staking_path"]
+                    ),
+                    nonce=catalyst_registration["nonce"],
```

Comment:
We've noticed here that it is possible to send numbers larger than uint64 max (18446744073709551615) and they simply go through to the Firmware, although the protobuf type is uint64. Cardano transaction signing then fails because the CBOR implementation can't (and probably shouldn't) handle larger numbers. This is true for all integers which aren't checked for upper bounds in the FW itself.

We are not sure if this is something we should be handling in trezorlib/FW or whether this should be handled somewhere else. Could you please provide some insight into this?

If a larger value is sent via Trezor Connect, FW receives uint64 max value instead of the value which has been sent.

Reply:
I'm OK with not handling this and letting the CBOR error bubble up as a FirmwareError or whatever is the final result. If you are worried about someone doing this by mistake, the place to check is here in trezorlib.

The firmware is currently not capable of enforcing protobuf numeric field sizes. I am not sure if that can be a problem anywhere; in the usual case, the final encoder should be using fixed size field and the failure will appear there.